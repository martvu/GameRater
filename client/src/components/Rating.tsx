import { useState } from 'react';

interface RatingProps {
  rating: number;
  numRatings?: number;
  disabled?: boolean;
  onChange?: (rating: number) => void;
}

/**
 * Rating component
 * @param {number} rating - The rating of the game
 * @param {number} numRatings - The number of ratings
 * @param {boolean} disabled - Whether the rating is disabled
 * @param {function} onChange - Callback function when rating is changed
 */
export default function Rating({ rating, numRatings }: RatingProps) {
  return (
    <>
      {/*        <!-- Component: Basic Rating --> */}
      <div className="flex flex-col items-center">
        {/*          <!-- Rating --> */}
        <StarRating rating={rating} disabled={true} />
        {/*          <!-- Helper text --> */}
        <span className="text-xs leading-6 text-muted-foreground">
          Based on {numRatings} user {numRatings === 1 ? 'rating' : 'ratings'}
        </span>
      </div>
      {/*        <!-- End Basic Rating --> */}
    </>
  );
}

/**
 * StarRating component
 * @param {number} rating - The rating of the game
 * @param {boolean} disabled - Whether the rating is disabled
 * @param {function} onChange - Callback function when rating is changed
 */
export const StarRating = ({ rating, disabled, onChange }: RatingProps) => {
  const [ratingValue, setRatingValue] = useState(rating || 0);
  const [hover, setHover] = useState(0);

  const handleClick = (value: number) => {
    if (disabled) {
      return;
    }
    if (onChange) {
      onChange(value);
    }
    setRatingValue(value);
  };
  return (
    <>
      <div className="flex items-center font-light">
        {[...Array(5)].map((_star, index) => {
          index += 1;
          return (
            <button
              type="button"
              key={index}
              className={
                index <= (hover || ratingValue)
                  ? 'text-red-400'
                  : 'text-muted-foreground'
              }
              onClick={() => handleClick(index)}
              onMouseEnter={() => setHover(index)}
              onMouseLeave={() => setHover(ratingValue)}
              onDoubleClick={() => {
                setRatingValue(0);
                setHover(0);
              }}
              disabled={disabled}
            >
              <span className="text-2xl font-light">
                {index <= ratingValue ? '★' : '☆'}{' '}
                {/* Filled star or empty star */}{' '}
              </span>
            </button>
          );
        })}
        <span className="ml-2 text-muted-foreground">{ratingValue}/5</span>
      </div>
    </>
  );
};
