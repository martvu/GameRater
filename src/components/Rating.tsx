import { useState } from 'react';

interface RatingProps {
  rating: number;
  numRatings?: number;
  disabled?: boolean;
  onChange?: (rating: number) => void;
}

export default function Rating({ rating, numRatings }: RatingProps) {
  return (
    <>
      {/*        <!-- Component: Basic Rating --> */}
      <div className="flex flex-col items-center gap-2">
        {/*          <!-- Rating --> */}
        <StarRating rating={rating} disabled={true}/>
        {/*          <!-- Helper text --> */}
        <span className="text-xs leading-6 text-muted-foreground">
          based on {numRatings} user ratings
        </span>
      </div>
      {/*        <!-- End Basic Rating --> */}
    </>
  );
}


export const StarRating = ({rating, disabled, onChange}: RatingProps) => {
  const [ratingValue, setRatingValue] = useState(rating || 0);
  const [hover, setHover] = useState(0);

  const handleClick = (value: number) => {
    if (disabled) {
      return;
    }
    if (onChange) {
      onChange(value);
    }
    setRatingValue(value);
  }
  return (
    <>
      <div className="flex items-center font-light">
        {[...Array(5)].map((_star, index) => {
          index += 1;
          return (
            <button
              type="button"
              key={index}
              className={(
                index <= (hover || ratingValue)
                  ? 'text-yellow-400'
                  : 'text-muted-foreground'
          )}
              onClick={() => handleClick(index)}
              onMouseEnter={() => setHover(index)}
              onMouseLeave={() => setHover(ratingValue)}
              onDoubleClick={() => {
                setRatingValue(0);
                setHover(0);
              }}
              disabled={disabled}
            >
              <span className="text-2xl font-light">
                {index <= rating ? '★' : '☆'} {/* Filled star or empty star */}{' '}
              </span>
            </button>
          );
        })}
        <span className='text-muted-foreground ml-2'>{rating}/5</span>
      </div>
      
    </>
  );
};

