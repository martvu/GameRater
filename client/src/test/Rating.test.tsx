import { describe, it, expect, vi } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Rating, { StarRating } from '@/components/Rating';

describe('Rating Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(<Rating rating={3} numRatings={100} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders star rating based on provided rating', () => {
    render(<Rating rating={3} numRatings={100} />);
    // Assuming the StarRating component displays the rating value as text
    expect(screen.getByText('3/5')).toBeInTheDocument();
  });

  it('displays the correct number of user ratings', () => {
    render(<Rating rating={4} numRatings={25} />);
    expect(screen.getByText('Based on 25 user ratings')).toBeInTheDocument();
  });

  it('handles singular form for one rating', () => {
    render(<Rating rating={5} numRatings={1} />);
    expect(screen.getByText('Based on 1 user rating')).toBeInTheDocument();
  });
});

describe('StarRating Component', () => {
  it('renders correctly with initial rating', () => {
    render(<StarRating rating={3} numRatings={100} disabled={false} />);
    expect(screen.getAllByText('★')).toHaveLength(3);
    expect(screen.getAllByText('☆')).toHaveLength(2);
  });

  it('calls onChange with the correct rating on click', async () => {
    const onChangeMock = vi.fn();
    render(<StarRating rating={2} onChange={onChangeMock} disabled={false} />);
    const thirdStar = screen.getAllByText('☆')[0];
    await userEvent.click(thirdStar);
    expect(onChangeMock).toHaveBeenCalledWith(3);
  });

  it('shows correct number of filled stars on hover', async () => {
    render(<StarRating rating={2} disabled={false} />);
    const fourthStar = screen.getAllByText('☆')[1];
    await userEvent.click(fourthStar);
    await waitFor(() => {
      expect(screen.getAllByText('★')).toHaveLength(4);
    });
  });

  it('does not allow interaction when disabled', async () => {
    render(<StarRating rating={3} disabled={true} />);
    const fourthStar = screen.getAllByText('☆')[0];
    await userEvent.click(fourthStar);
    expect(screen.getAllByText('★')).toHaveLength(3); // Rating should not change
  });

  it('resets rating on double click', async () => {
    render(<StarRating rating={3} disabled={false} />);
    const thirdStar = screen.getAllByText('★')[2];
    await userEvent.dblClick(thirdStar);
    expect(screen.getAllByText('☆')).toHaveLength(5); // All stars should be empty
  });

  it('reverts to original rating on mouse leave', async () => {
    render(<StarRating rating={2} disabled={false} />);

    const starToHover = screen.getAllByText('☆')[0];
    await userEvent.hover(starToHover);
    await userEvent.unhover(starToHover);

    expect(screen.getAllByText('★')).toHaveLength(2); // Check if it reverts to original rating
  });
});
