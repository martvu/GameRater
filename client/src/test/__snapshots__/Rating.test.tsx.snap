// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Rating Component > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-center"
  >
    <div
      class="flex items-center font-light"
    >
      <button
        aria-label="review-star-1"
        class="text-red-400"
        data-testid="review-star"
        disabled=""
        type="button"
      >
        <span
          class="text-2xl font-light"
        >
          ★  
        </span>
      </button>
      <button
        aria-label="review-star-2"
        class="text-red-400"
        data-testid="review-star"
        disabled=""
        type="button"
      >
        <span
          class="text-2xl font-light"
        >
          ★  
        </span>
      </button>
      <button
        aria-label="review-star-3"
        class="text-red-400"
        data-testid="review-star"
        disabled=""
        type="button"
      >
        <span
          class="text-2xl font-light"
        >
          ★  
        </span>
      </button>
      <button
        aria-label="review-star-4"
        class="text-muted-foreground"
        data-testid="review-star"
        disabled=""
        type="button"
      >
        <span
          class="text-2xl font-light"
        >
          ☆  
        </span>
      </button>
      <button
        aria-label="review-star-5"
        class="text-muted-foreground"
        data-testid="review-star"
        disabled=""
        type="button"
      >
        <span
          class="text-2xl font-light"
        >
          ☆  
        </span>
      </button>
      <span
        class="ml-2 text-muted-foreground"
      >
        3/5
      </span>
    </div>
    <span
      class="text-xs leading-6 text-muted-foreground"
    >
      Based on 100 user ratings
    </span>
  </div>
</DocumentFragment>
`;
